contracts/Orders.sol

```diff
@@ -0,0 +1,235 @@
+            order.dstAmount,
+            order.prevId,
+            order.nextId
+        );
+    }
+
+    function removeById(uint32 orderId) internal {
+        verifyCanRemoveOrderById(orderId);
+
+        // Disconnect order from list
+        Order storage order = orders[orderId];
+        orders[order.prevId].nextId = order.nextId;
+        orders[order.nextId].prevId = order.prevId;
+
+        order.prevId = 0;
+        order.nextId = 0;
+    }
+
+    // The updated order id is returned following the update.
+    function update(uint32 orderId, uint128 srcAmount, uint128 dstAmount)
+        internal
+        returns(uint32)
+    {
+        address maker = orders[orderId].maker;
+        removeById(orderId);
+        return add(maker, srcAmount, dstAmount);
+    }
+
+    // The updated order id is returned following the update.
+    function updateWithPositionHint(
+        uint32 orderId,
+        uint128 srcAmount,
+        uint128 dstAmount,
+        uint32 prevId
+    )
+        internal
+        returns(uint32)
+    {
+        address maker = orders[orderId].maker;
+        removeById(orderId);
+        return addAfterId(maker, srcAmount, dstAmount, prevId);
+    }
+
+    function allocateIds(uint32 howMany) internal returns(uint32) {
+        uint32 firstId = nextId;
+        nextId += howMany;
+        return firstId;
+    }
+
+    function verifyCanRemoveOrderById(uint32 orderId) private view {
+        require(orderId != BUY_HEAD_ID);
+
+        Order storage order = orders[orderId];
+
+        // Make sure such order exists in mapping.
+        require(order.prevId != 0 || order.nextId != 0);
+    }
+
+    function calculateOrderSortKey(uint128 srcAmount, uint128 dstAmount)
+        public
+        pure
+        returns(uint)
+    {
+        return dstAmount * PRECISION / srcAmount;
+    }
+
+    function findPrevOrderId(
+        uint128 srcAmount,
+        uint128 dstAmount,
+        uint32 startId
+    )
+        public
+        view
+        returns(uint32)
+    {
+        uint newOrderKey = calculateOrderSortKey(srcAmount, dstAmount);
+
+        // This is okay for the HEAD ids, as their prevId is their id.
+        // TODO: rewrite in a simpler way - going to the prev of the provided
+        //       first id is not that elegant.
+        uint32 currId = orders[startId].prevId;
+        Order storage curr = orders[currId];
+        while (curr.nextId != TAIL_ID) {
+            currId = curr.nextId;
+            curr = orders[currId];
+            uint key = calculateOrderSortKey(curr.srcAmount, curr.dstAmount);
+            if (newOrderKey > key) {
+                return curr.prevId;
+            }
+        }
+        return currId;
+    }
+
+    function validatePrevId(
+        uint128 srcAmount,
+        uint128 dstAmount,
+        uint32 prevId
+    )
+        private
+        view
+    {
+        // Make sure prev is not the tail.
+        require(prevId != TAIL_ID);
+
+        Order storage prev = orders[prevId];
```

Comment:
would add here option that runs this in memory and not storage pointer, and see gas report in truffle...